### DISCUSSION TITLE
API discussion / ideas

### DISCUSSION THREAD

**Discussion by Dead2** [zlib-ng/zlib-ng#852] [General]
Regarding possible API changes, I have been wondering whether the support for reading/writing gzip files is really needed. I want to keep gzip stream support, but I am not so sure about the support for actually doing the reading and writing of files itself. It seems like a lot of extra code for something that is not really specific to the compression algorithm, and often not used.

Is there any easy way to check what applications actually use those functions in an average linux distro?

**Comment by daxtens** [comment]
I don't think you'll be able to break API compatibility ... ever. At least not if you want distros to ever pick this up.

**Comment by Dead2** [comment]
There are several possible solutions to that, either to be parallel installable alongside stock zlib, or to have a compatibility-shim on top of zlib-ng to support legacy zlib applications.

**Comment by Dead2** [comment]
#### Zlib compatability

We could implement a ZLIB_COMPAT define that is optional and default off, that way we can separate the gzfile parts of the code and let the functions be deprecated with a scheduled future removal. Distros can then choose to enable the compat define for initial releases, and then disable it when they no longer see the need (giving upstreams time to change their code).

This gives us the opportunity to plan removal of old parts of the API that are not used much or are giving development/maintenance problems and at the same time letting application developers take the time they need to change their usage of zlib/zlib-ng and let it propagate into stable releases.

This could at the same time mostly replace Z_SOLO with the exception of malloc functionality, we would need to check whether that is actually needed for embedded systems or something.

**Comment by Dead2** [comment]
#### External buffers

I would also like to change the way we accept external buffers (in place of mallocing them ourselves), since that is a major blocker due to current applications providing a buffer that is too small for zlib-ng due to SSE alignment needs.

With the compat flag set we could for example discard the buffers if they are too small and malloc new ones anyways (or disable SSE usage when in compat mode, but that seems like a bad choice).

When the compat flag is not set then either remove the apis that accept external buffers, or add a method for the application to ask zlib-ng what buffer sizes it needs.

Nginx is one example of applications that provide their own buffers, and this can be a performance boost since it allows the workers to pre-allocate and possibly re-use the buffers (I have not looked into what they do, only speculations). Certain applications use non-standard malloc implementations to obtain some extra performance (TCMalloc and jemalloc are a few examples of often used malloc libraries).

So I think it would not be good to remove the support for external buffers, but something needs to be done to inform the application of the current zlib\* librarys needs. If we made a new simple function for this then we might be able to persuade stock zlib to copy it, meaning all applications that care for external buffers can transition to that method while retaining cross compatability. It lets us provide updated accurate sizes regardless of what zlib variant the application has been compiled with too. This is my preferred solution at the moment.
Btw, this is a problem for other zlib-compatible libraries as well, IPP for example but others as well. They would likely appreciate this addition.

**Comment by daxtens** [comment]
Hmm. So my perspective, doing this as a semi-official-work-ish project, is that I'd like to see it get taken up widely, and quickly. (I'm looking to add some PowerPC specific optimisations, but I'd like to get the code and test quality up first.)

I think in order to achieve wide uptake, we need to start with the "thin end of the wedge", and work up towards more radical changes. 

First, we should provide a byte-for-byte equivalent that is slightly faster: optimised CRCs, vectorised loops. I'm working on a big test suite that will help us establish byte-for-byte equivalence. I know it won't be a lot faster, but it will be a lot easier to get people to pick up a version that's absolutely identical and slightly faster.

Secondly, we provide flags to give complete API equivalence but with different size/speed tradeoffs: faster strategies with equivalent but not quite identical size ratios --- the new Quick and Medium strategies. Ideally we avoid different buffer sizes at this point.

After that, we can look at providing flags or something to deal with SSE buffer sizing. (I think we may actually be able to work around it... haven't looked yet)

Beyond that, we can look at providing a nicer API.

I think each stage is a wider 'wedge' away from the original zlib. As the wedge gets wider, it's harder and harder to convince people to pick it up and start using it. As such, I think we need to start by providing a faster and completely equivalent version, and then widening out to an improved version.

**Comment by Dead2** [comment]
While that is the safest way to go forward, I think that a clean break would be just as likely to be accepted into distros. If zlib-ng is paralell-installable along with zlib then applications that want it can make use of it. If the distro can also choose to install another zlib-ng in zlib-compatibility mode (as a zlib replacement) then they cover the applications that do not want to quickly add support for zlib-ng directly.

So for example: package zlib-ng provides zlib-ng.so and zlib-ng-compat provides zlib.so.
The compat package could be controlled through the 'alternatives' system that some distros have, such that the end-user can choose.

With this approach, byte-for-byte compatibility will not matter much, since a distro can install zlib-ng and let only certain applications use it, and when confidence rises, more will follow. And at some point zlib-ng in compat mode can replace the official zlib entirely.

From a distro view, outright replacing zlib will be a big risk to take, especially with the huge amount of changes we have made everywhere in the code. Going for a parallel installable lib would pretty much be allowed in everywhere and can even be backported to older distro releases.

I too want zlib-ng out in distros quickly, but I'd like it to be a fast version that has already been through most of the huge reconstructions. zlib-ng-compat is also an attractive target and could be byte-for-byte compatible if needed. I think both can be achieved at the same time by the same code (although with a bunch of ifdefs).

Working around the buffer size problem will likely not be possible without breaking the api OR breaking what the old api promises (no internal mallocing, only external buffers), but if you have any ideas feel free.

**Comment by mtl1979** [comment]
I think that avoiding malloc insize zlib makes sense only in embedded systems, where the available memory is limited... For other systems, it's pretty much pointless as the alternatives are allowing realloc'ing the buffer as long as the memory location doesn't change or returning with a buffer overflow error to the caller. Trying to work out the problem with misaligning-by-moving everytime when returning from assembler or intrinsic function is just going to slow down.

**Comment by CyberTech** [comment]
@mtl1979 , in reference to avoiding malloc inside zlib -- it's not just for embedded systems, it also affects those of us using zlib in kernel mode (regardless of platform).  For this case, it's ideal if the lib provides run-time configurable allocate/deallocate functions.

**Comment by mtl1979** [comment]
@CyberTech The problem with run-time configurable allocate/deallocate functions is that we need to add sanity tests for some platforms for cases that might not happen with library-provided allocate/deallocate functions. This becomes an issue with big-endian platforms where unaligned access is not allowed at all, and with some little-endian optimizations where some instructions don't allow unaligned access and crash or behave erradically (for example low bits are masked out or ignored), and the default alignment of the allocate/deallocate functions is not adequate.

**Comment by RJVB** [comment]
I didn't get the impression that that would be a generic feature that would have to be supported across the whole range of supported platforms.
But the easiest implementation would just provide a set of hooks to register user-provided (de)allocation routines, and exhaustive documentation of the requirements those routines have to satisfy.
If you're embedding the library in a kernel or firmware chances are that there are already appropriate functions available that you can tap in to.

**Comment by mtl1979** [comment]
I'll leave writing the documentation to @Dead2 but as I already mentioned elsewhere, main culprit is that due to processor limitations memory access in optimized code must be aligned to minimum of 16 byte boundary but stock malloc only guarantees 4 or 8 byte alignment depending on platform.
zlib-ng already has hooks to allocation and deallocation functions but some parts of the code do override them. That means we need to make the sanity checks in at least two places in zlib-ng code for any platform that has stricter requirements than stock GNU malloc.
